plugins/auth/src/components/ConsentPage/ConsentPage.tsx

```diff
@@ -0,0 +1,348 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { useCallback, useEffect, useState } from 'react';
+import { useParams } from 'react-router-dom';
+import {
+  Box,
+  Button,
+  Card,
+  CardContent,
+  CardActions,
+  Typography,
+  makeStyles,
+  Divider,
+} from '@material-ui/core';
+import { Alert } from '@material-ui/lab';
+import CheckCircleIcon from '@material-ui/icons/CheckCircle';
+import CancelIcon from '@material-ui/icons/Cancel';
+import AppsIcon from '@material-ui/icons/Apps';
+import WarningIcon from '@material-ui/icons/Warning';
+import {
+  Header,
+  Page,
+  Content,
+  Progress,
+  EmptyState,
+  ResponseErrorPanel,
+} from '@backstage/core-components';
+import {
+  alertApiRef,
+  useApi,
+  fetchApiRef,
+  discoveryApiRef,
+} from '@backstage/core-plugin-api';
+import { isError } from '@backstage/errors';
+
+const useStyles = makeStyles(theme => ({
+  authCard: {
+    maxWidth: 600,
+    margin: '0 auto',
+    marginTop: theme.spacing(4),
+  },
+  appHeader: {
+    display: 'flex',
+    alignItems: 'center',
+    marginBottom: theme.spacing(2),
+  },
+  appIcon: {
+    marginRight: theme.spacing(2),
+    fontSize: 40,
+  },
+  appName: {
+    fontSize: '1.5rem',
+    fontWeight: 'bold',
+  },
+  securityWarning: {
+    margin: theme.spacing(2, 0),
+  },
+  buttonContainer: {
+    display: 'flex',
+    justifyContent: 'space-between',
+    gap: theme.spacing(2),
+    padding: theme.spacing(2),
+  },
+  callbackUrl: {
+    fontFamily: 'monospace',
+    backgroundColor: theme.palette.background.default,
+    padding: theme.spacing(1),
+    borderRadius: theme.shape.borderRadius,
+    wordBreak: 'break-all',
+    fontSize: '0.875rem',
+  },
+  scopeList: {
+    backgroundColor: theme.palette.background.default,
+    borderRadius: theme.shape.borderRadius,
+    padding: theme.spacing(1),
+  },
+}));
+
+interface Session {
+  id: string;
+  clientName?: string;
+  clientId: string;
+  redirectUri: string;
+  scopes?: string[];
+  responseType?: string;
+  state?: string;
+  nonce?: string;
+  codeChallenge?: string;
+  codeChallengeMethod?: string;
+  expiresAt?: string;
+}
+
+export const ConsentPage = () => {
+  const classes = useStyles();
+  const { sessionId } = useParams<{ sessionId: string }>();
+  const alertApi = useApi(alertApiRef);
+  const fetchApi = useApi(fetchApiRef);
+  const discoveryApi = useApi(discoveryApiRef);
+  const [session, setSession] = useState<Session | null>(null);
+  const [loading, setLoading] = useState(true);
+  const [submitting, setSubmitting] = useState(false);
+  const [error, setError] = useState<string | null>(null);
+  const [completed, setCompleted] = useState<
+    | {
+        action: 'approve' | 'reject';
+      }
+    | undefined
+  >(undefined);
```

Comment:
Spontaneously feel like all of these could have been a single state with the type `{ state: 'loading' } | { state: 'error'; error: string } | ...` to clarify things a bit.

Or at least, capturing all of the loading and error and session state, in a single `useAsync` which already handles all of that jazz